File: python/data_structures/linkedlist/doubly_linked_list/doubly_linked_list_scratch.py
class Node:
    def __init__(self, value):
        self.value = value
        self.prev = None  # pointer to previous node
        self.next = None  # pointer to next node

    def __repr__(self):
        return f"Node({self.value!r})"


class DoublyLinkedList:

    def __init__(self):
        self.head = None
        self.tail = None
        self._size = 0

    # ---------- Helpers ----------
    def is_empty(self):
        return self._size == 0

    def size(self):
        return self._size

    def __len__(self):
        return self._size

    def __iter__(self):
        current = self.head
        while current is not None:
            yield current.value
            current = current.next

    def __repr__(self):
        values = ", ".join(repr(v) for v in self)
        return f"DoublyLinkedList([{values}])"

    # ---------- Core operations ----------
    def push_front(self, x):
        new_node = Node(x)

        if self.is_empty():
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node

        self._size += 1
        return new_node

    def push_back(self, x):
        new_node = Node(x)

        if self.is_empty():
            # list is empty: head and tail become the new node
            self.head = self.tail = new_node
        else:
            # link new_node after current tail
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node

        self._size += 1
        return new_node

    def insert_after(self, node, x):
        if node is None:
            raise ValueError("node must not be None")

        if node is self.tail:
            return self.push_back(x)

        new_node = Node(x)

        # link new_node between node and node.next
        successor = node.next
        new_node.prev = node
        new_node.next = successor

        node.next = new_node
        successor.prev = new_node

        self._size += 1
        return new_node

    def insert_before(self, node, x):
        if node is None:
            raise ValueError("node must not be None")

        if node is self.head:
            return self.push_front(x)

        new_node = Node(x)

        predecessor = node.prev
        new_node.next = node
        new_node.prev = predecessor

        predecessor.next = new_node
        node.prev = new_node

        self._size += 1
        return new_node

    def pop_front(self):
        if self.is_empty():
            raise IndexError("pop_front from empty list")

        node = self.head
        value = node.value

        if self.head is self.tail:
            # only one element
            self.head = self.tail = None
        else:
            # move head forward
            self.head = self.head.next
            self.head.prev = None

        self._size -= 1
        return value

    def pop_back(self):
        if self.is_empty():
            raise IndexError("pop_back from empty list")

        node = self.tail
        value = node.value

        if self.head is self.tail:
            # only one element
            self.head = self.tail = None
        else:
            # move tail backward
            self.tail = self.tail.prev
            self.tail.next = None

        self._size -= 1
        return value

    def remove(self, node):
        if node is None:
            raise ValueError("node must not be None")

        if self.is_empty():
            raise IndexError("remove from empty list")

        # If it's the only node
        if node is self.head and node is self.tail:
            self.head = self.tail = None

        # If it's the head (but not the only node)
        elif node is self.head:
            self.head = node.next
            self.head.prev = None

        # If it's the tail (but not the only node)
        elif node is self.tail:
            self.tail = node.prev
            self.tail.next = None

        # Node is in the middle
        else:
            prev_node = node.prev
            next_node = node.next
            prev_node.next = next_node
            next_node.prev = prev_node

        # help garbage collector
        node.prev = None
        node.next = None

        self._size -= 1
        return node.value

    def find(self, value):
        current = self.head
        while current is not None:
            if current.value == value:
                return current
            current = current.next
        return None

File: java/data_structures/linkedlist/doubly_linked_list/DoublyLinkedListScratch.java
package data_structures.linkedlist.doubly_linked_list;

import java.util.NoSuchElementException;

public class DoublyLinkedListScratch<T> {

    public static class Node<T> {
        T value;
        Node<T> prev;
        Node<T> next;

        public Node(T value) {
            this.value = value;
        }

        public T getValue() {
            return value;
        }

        public Node<T> getPrev() {
            return prev;
        }

        public Node<T> getNext() {
            return next;
        }

        @Override
        public String toString() {
            return "Node(" + String.valueOf(value) + ")";
        }
    }

    private Node<T> head;
    private Node<T> tail;
    private int sizeValue;

    public DoublyLinkedListScratch() {
        this.head = null;
        this.tail = null;
        this.sizeValue = 0;
    }

    public boolean isEmpty() {
        return sizeValue == 0;
    }

    public int size() {
        return sizeValue;
    }

    private Node<T> makeNode(T value) {
        return new Node<>(value);
    }

    public Node<T> getHead() {
        return head;
    }

    public Node<T> getTail() {
        return tail;
    }

    
    public Node<T> pushFront(T value) {
        Node<T> n = makeNode(value);

        n.next = head;

        if (head != null) {
            head.prev = n;
        } else {
            tail = n;
        }

        head = n;
        sizeValue++;
        return n;
    }

    public Node<T> pushBack(T value) {
        Node<T> n = makeNode(value);

        if (tail == null) {
            head = tail = n;
        } else {
            n.prev = tail;
            tail.next = n;
            tail = n;
        }

        sizeValue++;
        return n;
    }

    public Node<T> insertAfter(Node<T> node, T value) {
        if (node == null) {
            throw new IllegalArgumentException("insertAfter: node cannot be null");
        }

        if (node == tail) {
            return pushBack(value);
        }

        Node<T> n = makeNode(value);
        Node<T> successor = node.next;

        n.prev = node;
        n.next = successor;

        node.next = n;
        if (successor != null) {
            successor.prev = n;
        }

        sizeValue++;
        return n;
    }

    public Node<T> insertBefore(Node<T> node, T value) {
        if (node == null) {
            throw new IllegalArgumentException("insertBefore: node cannot be null");
        }

        if (node == head) {
            return pushFront(value);
        }

        Node<T> n = makeNode(value);
        Node<T> predecessor = node.prev;

        n.next = node;
        n.prev = predecessor;

        if (predecessor != null) {
            predecessor.next = n;
        }
        node.prev = n;

        sizeValue++;
        return n;
    }

    public T popFront() {
        if (isEmpty()) {
            throw new NoSuchElementException("popFront from empty list");
        }

        Node<T> node = head;
        T value = node.value;

        if (head == tail) {
            head = tail = null;
        } else {
            head = head.next;
            head.prev = null;
        }

        sizeValue--;
        return value;
    }

    public T popBack() {
        if (isEmpty()) {
            throw new NoSuchElementException("popBack from empty list");
        }

        Node<T> node = tail;
        T value = node.value;

        if (head == tail) {
            head = tail = null;
        } else {
            tail = tail.prev;
            tail.next = null;
        }

        sizeValue--;
        return value;
    }

    public T remove(Node<T> node) {
        if (node == null) {
            throw new IllegalArgumentException("remove: node cannot be null");
        }

        if (isEmpty()) {
            throw new NoSuchElementException("remove from empty list");
        }

        if (node == head && node == tail) {
            head = tail = null;
        }
        else if (node == head) {
            head = node.next;
            head.prev = null;
        }
        else if (node == tail) {
            tail = node.prev;
            tail.next = null;
        }
        else {
            Node<T> prevNode = node.prev;
            Node<T> nextNode = node.next;
            prevNode.next = nextNode;
            nextNode.prev = prevNode;
        }

        node.prev = null;
        node.next = null;

        sizeValue--;
        return node.value;
    }

    public Node<T> find(T value) {
        Node<T> current = head;
        while (current != null) {
            if ((current.value == null && value == null) ||
                (current.value != null && current.value.equals(value))) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("DoublyLinkedListScratch[");
        Node<T> current = head;
        while (current != null) {
            sb.append(String.valueOf(current.value));
            if (current.next != null) {
                sb.append(", ");
            }
            current = current.next;
        }
        sb.append("]");
        return sb.toString();
    }
}
